src/main/java/io/stargate/sgv2/jsonapi/service/operation/collections/CollectionReadOperation.java

```diff
@@ -389,21 +390,37 @@ default Uni<CountResponse> countDocumentsByKey(
       QueryExecutor queryExecutor,
       SimpleStatement simpleStatement) {
     AtomicLong counter = new AtomicLong();
-    final CompletionStage<AsyncResultSet> async =
-        queryExecutor
-            .executeCount(dataApiRequestInfo, simpleStatement)
-            .whenComplete(
-                (rs, error) -> {
-                  getCount(rs, error, counter);
-                });
 
-    return Uni.createFrom()
-        .completionStage(async)
+    return Multi.createBy()
+        .repeating()
+        .uni(
+            () -> new AtomicReference<AsyncResultSet>(null), // the state passed to the producer
+            stateRef -> {
+              Uni<AsyncResultSet> result =
+                  stateRef.get() == null
+                      ? Uni.createFrom()
+                          .completionStage(
+                              queryExecutor.executeCount(
+                                  dataApiRequestInfo,
+                                  simpleStatement)) // AsyncResultSet is null so first page
+                      : Uni.createFrom()
+                          .completionStage(stateRef.get().fetchNextPage()); // After first page
+              // returning result for looping, this is remembering the rows as we page through them
+              return result
+                  .onItem()
+                  .invoke(
+                      rs -> {
+                        counter.addAndGet(rs.remaining());
+                        stateRef.set(rs);
+                      });
+            })
+        // Documents read until pageState available, max records read is deleteLimit + 1 TODO
+        // COMMENTS
+        .whilst(AsyncPagingIterable::hasMorePages)
```

Comment:
Quick question: where is the "max count" checked? Looks like this would read through all entries and not stop at specific count? (I am probably missing something)

Reply:
this looks like the code we came up with for in memory sorting for tables

Reply:
Ah. Yes, since that's within `SimpleStatement` being passed.